src/clsourcedependencies_parser.cc

```diff
@@ -0,0 +1,106 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "clsourcedependencies_parser.h"
+
+#include "rapidjson/document.h"
+
+#include "clparser.h"
+#include "metrics.h"
+#include "util.h"
+
+using namespace rapidjson;
+using namespace std;
+
+bool CLSourceDependenciesParser::Parse(const string& content, string* err) {
+  METRIC_RECORD("CLSourceDependenciesParser::Parse");
+
+  Document doc;
+  if (doc.Parse(content.c_str()).HasParseError()) {
+    *err = string("sourceDependencies parse error");
+    return false;
+  }
+
+  if (!doc.IsObject()) {
+    *err = string("sourceDependencies root must be an object");
+    return false;
+  }
+
+  if (!doc.HasMember("Version")) {
+    *err = string("sourceDependencies missing version");
+    return false;
+  }
+
+  const Value& version = doc["Version"];
+  if (!version.IsString()) {
+    *err = string("sourceDependencies version must be a string");
+    return false;
+  }
+
+  const char* version_str = version.GetString();
+  if (strncmp(version_str, "1.", 2)) {
+    *err = string("expected sourceDependencies version 1.x but found ") +
+           version_str;
+    return false;
+  }
+
+  if (!doc.HasMember("Data")) {
+    *err = string("sourceDependencies missing data");
+    return false;
+  }
+
+  const Value& data = doc["Data"];
+  if (!data.IsObject()) {
+    *err = string("sourceDependencies data must be an object");
+    return false;
+  }
+
+  if (!data.HasMember("Includes")) {
+    *err = string("sourceDependencies data missing includes");
+    return false;
+  }
+
+  const Value& includes = data["Includes"];
+  if (!includes.IsArray()) {
+    *err = string("sourceDependencies includes must be an array");
+    return false;
+  }
+
+  for (Value::ConstValueIterator i = includes.Begin(); i != includes.End();
+       ++i) {
+    if (!i->IsString()) {
+      *err = string("sourceDependencies include path must be a string");
+      return false;
+    }
+
+    const char* include_c_str = i->GetString();
+#ifdef _WIN32
+    // JSON is encoded in UTF-8, but Ninja on Windows uses the ANSI APIs.
+    // Attempt to convert to the active code page. This could fail if the user
+    // has characters outside the code page in their paths, but there's not much
+    // we can do about that without more fundamental changes to Ninja.
```

Comment:
I recommend depending on this PR instead: https://github.com/ninja-build/ninja/pull/1671

Reply:
#1671 looks like it's stuck in limbo. @jhasse is there any chance it will be merged soon?

Reply:
It will be merged before this PR, so you can depend on it.

Reply:
I removed the UTF8 to ASCII conversion. This PR will benefit from #1671, but it's not a required prerequisite. Without it, ASCII paths will work correctly but non-ASCII paths may have issues depending on the active code page. This is not a new issue #824.